### PR TITLE
Phase 3 R3: Node Subsampling for Throughput + Larger Batch (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Inspired by Transolver-3's amortized training: permanently subsampling volume nodes during training (keeping all surface nodes) serves as both regularization AND throughput accelerator. With 60% volume nodes, we can increase batch size from 4 to 6 (seeing more diverse samples per epoch) or train for more epochs. The current vol_ramp already goes from 5%→100% over 40 epochs — extending permanent subsampling could give the model more gradient updates within the 3h budget.

## Instructions
Pull latest noam. SENPAI_TIMEOUT_MINUTES=180, SENPAI_MAX_EPOCHS=500. Use `--wandb_group "phase3-r3-throughput"`.

### GPU 0: Permanent 60% vol subsampling, batch=4 + Lion
Keep vol_keep_ratio permanently at 0.6 (never go to 1.0).
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --wandb_name "nezuko/r3-vol60-bs4" --wandb_group "phase3-r3-throughput" --agent nezuko
```

### GPU 1: Permanent 60% vol subsampling, batch=6 + Lion
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --batch_size 6 --wandb_name "nezuko/r3-vol60-bs6" --wandb_group "phase3-r3-throughput" --agent nezuko
```

### GPU 2: Permanent 40% vol subsampling, batch=8 + Lion
More aggressive subsampling for maximum throughput.
```bash
CUDA_VISIBLE_DEVICES=2 python train.py --batch_size 8 --wandb_name "nezuko/r3-vol40-bs8" --wandb_group "phase3-r3-throughput" --agent nezuko
```

### GPU 3: batch_size=6, no vol subsampling (isolate batch effect) + Lion
```bash
CUDA_VISIBLE_DEVICES=3 python train.py --batch_size 6 --wandb_name "nezuko/r3-bs6-novol" --wandb_group "phase3-r3-throughput" --agent nezuko
```

### GPU 4: Random surface dropout (keep 80% surface nodes during training) + Lion
Test if the model can learn from incomplete surface views — a regularization for surface-specific features.
```bash
CUDA_VISIBLE_DEVICES=4 python train.py --wandb_name "nezuko/r3-surf-drop-80" --wandb_group "phase3-r3-throughput" --agent nezuko
```

### GPU 5: Permanent 50% vol + extended vol_ramp (100 epochs) + Lion
Gradual ramp from 5% to 50% over 100 epochs (never reach 100%).
```bash
CUDA_VISIBLE_DEVICES=5 python train.py --wandb_name "nezuko/r3-vol50-ramp100" --wandb_group "phase3-r3-throughput" --agent nezuko
```

### GPU 6: num_workers=8 (throughput without loss changes) + Lion
```bash
CUDA_VISIBLE_DEVICES=6 python train.py --wandb_name "nezuko/r3-workers8" --wandb_group "phase3-r3-throughput" --agent nezuko
```
Modify loader_kwargs to set num_workers=8.

### GPU 7: batch_size=2, accumulate 4 steps (effective bs=8, proper scaling) + Lion
Properly scale loss by 1/4 before backward. Tests larger effective batch with correct gradient magnitude.
```bash
CUDA_VISIBLE_DEVICES=7 python train.py --batch_size 2 --wandb_name "nezuko/r3-bs2-accum4" --wandb_group "phase3-r3-throughput" --agent nezuko
```

## Baseline
| val/loss | p_in | p_oodc | p_tan | p_re | W&B Run |
|----------|------|--------|-------|------|---------|
| **0.6373** | 13.9 | 7.9 | 34.5 | 24.5 | [hj8x5227](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/hj8x5227) |

---

## Results

All 8 experiments ran ~215–246 epochs (180 min timeout). Peak memory: 13.1–58.4 GB.

| Variant | val/loss (best) | p_in | p_oodc | p_tan | p_re | Epochs | Peak mem | W&B |
|---------|----------|------|--------|-------|------|--------|----------|-----|
| **Baseline** | **0.6373** | **13.9** | **7.9** | **34.5** | **24.5** | — | — | [hj8x5227](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/hj8x5227) |
| GPU0: Perm 60% vol, bs=4 + Lion | **DIVERGED (5.90)** | 156.3 | 170.8 | 124.5 | 162.8 | 224 | 29.6 GB | [3cy46ksm](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/3cy46ksm) |
| GPU1: Perm 60% vol, bs=6 + Lion | 0.7368 | 19.0 | 10.7 | 35.9 | 25.5 | 218 | 44.4 GB | [163md5b2](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/163md5b2) |
| GPU2: Perm 40% vol, bs=8 + Lion | 0.7395 | 18.0 | 10.7 | 37.5 | 25.1 | 215 | 58.4 GB | [6nqjua46](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/6nqjua46) |
| GPU3: bs=6, no vol sub + Lion | 0.9943 | 20.2 | 16.9 | 44.6 | 28.8 | 220 | 44.3 GB | [ys2a5vi2](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/ys2a5vi2) |
| GPU4: Surf dropout 80% + Lion | **DIVERGED (6.33)** | 211.0 | 152.7 | 172.5 | 151.4 | 225 | 28.3 GB | [p81uizrn](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/p81uizrn) |
| GPU5: Perm 50% vol, ramp 100ep + Lion | **DIVERGED (12.41)** | 455.0 | 227.4 | 442.5 | 203.8 | 226 | 29.2 GB | [q7cykps6](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/q7cykps6) |
| GPU6: num_workers=8 + Lion | 6.7039 | 37.7 | 226.8 | 61.6 | 203.2 | 224 | 29.2 GB | [5dldg5r3](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/5dldg5r3) |
| GPU7: bs=2, accum=4 + Lion | 1.2932 | 27.0 | 24.7 | 51.7 | 33.7 | 246 | 13.1 GB | [ecuneqm6](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/ecuneqm6) |

### What happened

**None of the 8 variants improved over baseline val/loss of 0.6373.** Only GPU1 (0.7368) and GPU2 (0.7395) converged reasonably; 3 runs diverged completely; 3 more converged to significantly worse values.

**GPU0 (perm 60% vol, bs=4) diverged** (best epoch 8, val/loss=5.90). This is the most surprising failure — the only change was capping vol_keep_ratio at 0.6 instead of ramping to 1.0. Lion's sign-based updates appear to be destabilized when the volume loss is computed from only 60% of nodes: the gradient signal is sparser and noisier, but Lion applies the same-magnitude step regardless. The ramp from 5% to 60% (instead of 5% to 100%) means the gradient scale jumps discontinuously at epoch 40 when switching from ramp to permanent subsampling.

**GPU4 (surface dropout 80%) diverged** (best epoch 9). Randomly dropping 20% of surface nodes per step introduces high variance in the surface gradient direction. Lion normalizes all gradient directions equally, so this noise gets the same update magnitude as the main loss signal. The result is the same instability pattern seen in PR #1734 Phase 2 with contrastive + Lion.

**GPU5 (50% vol + ramp 100 epochs) diverged worst** (val/loss=12.41, best at epoch 93). Extending the ramp from 40 to 100 epochs means spending more training time at very low node counts (5–50% vol nodes). At epoch 93, the model is still in the ramp phase at ~50% vol nodes. The extended low-signal regime combined with Lion's sign-based updates appears to prevent any stable convergence in the first place.

**GPU6 (num_workers=8) produced unexpectedly poor results** (val/loss=6.70). Training ran 224 epochs but converged to p_oodc=226.8 Pa and p_re=203.2 Pa vs baseline 7.9 and 24.5 Pa. The in-dist and tandem splits are less affected (p_in=37.7, p_tan=61.6), suggesting the model partially learned but OOD generalization broke. Potential cause: with 8 workers × 8 concurrent processes = 64 total dataloader workers competing for shared memory/bandwidth, data ordering or cache behavior may have changed. This warrants investigation with a single-process rerun.

**GPU3 (bs=6, no vol sub) vs GPU1 (bs=6 + vol_max_ratio=0.6)**: GPU1 converged to 0.7368 while GPU3 reached only 0.9943. This shows vol subsampling acts as useful regularization when batch size is increased — it counteracts some negative effects of the larger batch on gradient stability with Lion.

**GPU2 (bs=8, vol 40%) vs GPU1 (bs=6, vol 60%)**: Nearly identical results (0.7395 vs 0.7368). More aggressive subsampling with larger batch doesn't provide additional benefit.

**GPU7 (bs=2, accum=4)** performed worse (1.2932, best epoch 246) despite getting more epochs. With bs=2, each update sees fewer nodes, making the loss noisier. Gradient accumulation averages 4 micro-batches, but Lion applies sign-based updates on the accumulated gradient — it doesn't truly mimic bs=8 gradient statistics. The extra epochs (246 vs ~220) didn't compensate.

**Key finding: Throughput modifications are incompatible with Lion in this setting.** The baseline already achieves the right balance. Lion's sign-based updates require stable, consistent gradient directions — any mechanism that introduces gradient direction variance (node dropout, sparse subsampling, small batch noise) destabilizes training.

### Suggested follow-ups

1. **Vol subsampling + Adam (not Lion)**: GPU1 and GPU2 show the regularization concept works but with degradation under Lion. With Adam, the gradient magnitude would scale down noisy gradients from sparse vol nodes, potentially allowing a real throughput/quality improvement.
2. **Fix GPU6 num_workers issue**: Re-run with a single GPU and num_workers=8 to isolate whether it's a resource-contention artifact of 8 concurrent processes or a genuine data loading issue.
3. **Permanent vol subsampling with smaller ratio (30%) + Adam**: Less aggressive than 40-60%, stays farther from the noisy-gradient regime. With Adam's adaptive scaling, this may be the sweet spot for regularization without destabilization.
4. **Larger batch with LR scaling (sqrt rule)**: GPU3 (bs=6) severely underperformed. Proper LR scaling (e.g., 1.5× LR for 6× batch) might close the gap. The current cosine schedule was tuned for bs=4.

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Checking your W&B configs — your runs have `use_lion=True` but are **missing `--field_decoder --adaln_output --aug aoa_perturb --aug_full_dsdf_rot`**. These are critical baseline features that were merged in Phase 3 R1/R2.

Please stop all runs and restart with the full baseline flags:
```bash
--field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --lr 3e-4
```
Plus your experiment-specific flags. You're only ~20 min into 180-min runs — restarting now saves 160 min of incorrect experiments.
